internal/server/analytics/clickhouse/mutation.go

```diff
@@ -0,0 +1,38 @@
+package clickhouse
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.flipt.io/flipt/internal/server/analytics"
+)
+
+// IncrementFlagEvaluationCounts does a batch insert of flag values from evaluations.
+func (c *Client) IncrementFlagEvaluationCounts(ctx context.Context, responses []*analytics.EvaluationResponse) error {
+	if len(responses) < 1 {
+		return nil
+	}
+
+	valuePlaceHolders := make([]string, 0, len(responses))
+	valueArgs := make([]interface{}, 0, len(responses)*8)
+
+	for _, response := range responses {
+		valuePlaceHolders = append(valuePlaceHolders, "(toDateTime(?, 'UTC'),?,?,?,?,?,?,?)")
+		valueArgs = append(valueArgs, response.Timestamp.Format(time.DateTime))
+		valueArgs = append(valueArgs, counterAnalyticsName)
+		valueArgs = append(valueArgs, response.NamespaceKey)
+		valueArgs = append(valueArgs, response.FlagKey)
+		valueArgs = append(valueArgs, response.FlagType)
+		valueArgs = append(valueArgs, response.Reason)
+		valueArgs = append(valueArgs, response.Match)
+		valueArgs = append(valueArgs, 1)
+	}
+
+	//nolint:gosec
+	stmt := fmt.Sprintf("INSERT INTO %s VALUES %s", counterAnalyticsTable, strings.Join(valuePlaceHolders, ","))
```

Comment:
Just a passing thought, nothing to really do for now.
TL;DR something to keep an eye on r.e. performance.

I imagine constantly rebuilding this query to be quite wasteful, particularly under load. I remember you saying their native client is more performant for inserts, so maybe we just don't sweat this until we eventually move to their other client. However, one way to mitigate this might be cache the insert query based on buffer size. To try and amortize the cost and avoid constantly reallocating this query.

I wonder if CH has a better way to prepare this query, so that the placeholders don't have to be repeated in this way too. That might be something their own client takes advantage of.

Aside: I looked at this first, because of the injection problem, and the `nosec` there made suspicious too. But I see you're building the placeholder query with this. Which itself is just a series of palceholders, no user inputs, so it is safe. Then you're passing the values via the placeholder on exec, which is good. So I can see why you step around the linting. (I am assuming the noline rule is upset here because of fmt stringing a SQL ).

Reply:
@GeorgeMac Yeah exactly! I tried to see if there was a better way to prepare this query. But you are right, we can look into that more native client.

I can place an issue in the project to do that. We can see if it has more features that is better suited for inserts.